src/librustc_lint/builtin.rs

```diff
@@ -1307,3 +1307,36 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnionsWithDropFields {
         }
     }
 }
+
+/// Lint for items marked `pub` that aren't exported to other crates
+pub struct UnexportedPub;
+
+declare_lint! {
+    UNEXPORTED_PUB,
```

Comment:
I'd actually rename the lint into `unreachable_pub` (+ update all the comments, etc).